File: static/js/blackjack.js
$(document).ready(function () {
    const $dealerCards = $('#dealer-cards');
    const $dealerScore = $('#dealer-score');
    const $playerHands = $('#player-hands');
    const $playerCardsContainer = $('#player-cards-container');
    const $playerCardsContainer2 = $('#player-cards-container-2');
    const $dealerCardsContainer = $('#dealer-cards-container');
    const $playerScores = $('#player-scores');
    const $result = $('#result');
    const $win = $('#win');
    const $hit = $('#hit');
    const $hit2 = $('#hit2');
    const $double = $('#double');
    const $split = $('#split');
    const $stand = $('#stand');
    const $stand2 = $('#stand2');
    const $start = $('#start');

    const cardShortNames = {
        'Туз ♠ пики': 'A♠️',
        '2 ♠ пики': '2♠️',
        '3 ♠ пики': '3♠️',
        '4 ♠ пики': '4♠️',
        '5 ♠ пики': '5♠️',
        '6 ♠ пики': '6♠️',
        '7 ♠ пики': '7♠️',
        '8 ♠ пики': '8♠️',
        '9 ♠ пики': '9♠️',
        '10 ♠ пики': '10♠️',
        'Валет ♠ пики': 'J♠️',
        'Дама ♠ пики': 'Q♠️',
        'Король ♠ пики': 'K♠️',
        'Туз ♥ черви': 'A♥️',
        '2 ♥ черви': '2♥️',
        '3 ♥ черви': '3♥️',
        '4 ♥ черви': '4♥️',
        '5 ♥ черви': '5♥️',
        '6 ♥ черви': '6♥️',
        '7 ♥ черви': '7♥️',
        '8 ♥ черви': '8♥️',
        '9 ♥ черви': '9♥️',
        '10 ♥ черви': '10♥️',
        'Валет ♥ черви': 'J♥️',
        'Дама ♥ черви': 'Q♥️',
        'Король ♥ черви': 'K♥️',
        'Туз ♦ бубны': 'A♦️',
        '2 ♦ бубны': '2♦️',
        '3 ♦ бубны': '3♦️',
        '4 ♦ бубны': '4♦️',
        '5 ♦ бубны': '5♦️',
        '6 ♦ бубны': '6♦️',
        '7 ♦ бубны': '7♦️',
        '8 ♦ бубны': '8♦️',
        '9 ♦ бубны': '9♦️',
        '10 ♦ бубны': '10♦️',
        'Валет ♦ бубны': 'J♦️',
        'Дама ♦ бубны': 'Q♦️',
        'Король ♦ бубны': 'K♦️',
        'Туз ♣ трефы': 'A♣️',
        '2 ♣ трефы': '2♣️',
        '3 ♣ трефы': '3♣️',
        '4 ♣ трефы': '4♣️',
        '5 ♣ трефы': '5♣️',
        '6 ♣ трефы': '6♣️',
        '7 ♣ трефы': '7♣️',
        '8 ♣ трефы': '8♣️',
        '9 ♣ трефы': '9♣️',
        '10 ♣ трефы': '10♣️',
        'Валет ♣ трефы': 'J♣️',
        'Дама ♣ трефы': 'Q♣️',
        'Король ♣ трефы': 'K♣️'
    };

    let existingPlayerCards = [];
    let existingDealerCards = [];
    const playerCardCounts = {};
    const dealerCardCounts = {};

    function addCardToArray(cardKey, card, cardArray, cardCounts) {
        if (!cardCounts[cardKey]) {
            cardCounts[cardKey] = 0;
        }
        cardCounts[cardKey]++;

        cardArray.push({key: cardKey, card: card, index: cardCounts[cardKey]});
    }

    function handlePlayerWin() {
        confetti({
            particleCount: 200,
            angle: 60,
            spread: 70,
            origin: {x: -0.1, y: 0.7}
        });

        confetti({
            particleCount: 200,
            angle: 120,
            spread: 70,
            origin: {x: 1.1, y: 0.7}
        });
    }

    function handlePlayerWinSmall() {
        confetti({
            particleCount: 150,
            angle: 60,
            spread: 50,
            origin: {x: -0.1, y: 0.9}
        });

        confetti({
            particleCount: 150,
            angle: 120,
            spread: 50,
            origin: {x: 1.1, y: 0.9}
        });
    }


    function updateGame(data) {
        if (!data.show_dealer_cards) {
            $dealerCards.html(data.dealer_cards.join(', '));
            $dealerScore.html('<b>Сумма: </b>' + data.dealer_score);
        } else {
            $dealerCards.html(data.dealer_cards[0] + ', *');
            $dealerScore.html('<b>Сумма:</b> *');
        }

        $playerHands.html(data.player_hands.map((hand, index) =>
            `<b>Рука ${index + 1}:</b> ${hand.join(', ')}`).join('<br>'));
        $playerScores.html('<b>Суммы:</b> ' + data.player_scores.join(', '));

        data.player_hands.forEach((hand, handIndex) => {
            const $targetContainer = handIndex === 0 ? $playerCardsContainer : $('#player-cards-container-2');

            hand.forEach((card, cardIndex) => {
                const cardKey = `${handIndex}-${cardIndex}`;

                if (!existingPlayerCards.some(c => c.key === cardKey)) {
                    addCardToArray(cardKey, card, existingPlayerCards, playerCardCounts);

                    const shortName = cardShortNames[card] || card;
                    const $cardDiv = $('<div>').addClass('card');
                    const $topLeft = $('<div>').addClass('top-left').text(shortName);
                    const $bottomRight = $('<div>').addClass('bottom-right').text(shortName);
                    if (shortName.includes('♠️') || shortName.includes('♣️')) {
                        $topLeft.css('color', 'black');
                        $bottomRight.css('color', 'black');
                    } else if (shortName.includes('♥️') || shortName.includes('♦️')) {
                        $topLeft.css('color', 'red');
                        $bottomRight.css('color', 'red');
                    }

                    $cardDiv.append($topLeft, $bottomRight);
                    setTimeout(() => {
                        $targetContainer.append($cardDiv);
                        setTimeout(() => {
                            $cardDiv.addClass('animate');
                        }, 50);
                    }, cardIndex * 150);
                }
            });
        });

        data.dealer_cards.forEach((card, cardIndex) => {
            const cardKey = `dealer-${cardIndex}`;
            if (!existingDealerCards.some(c => c.key === cardKey)) {
                addCardToArray(cardKey, card, existingDealerCards, dealerCardCounts);
                const shortName = cardShortNames[card] || card;
                const $cardDiv = $('<div>').addClass('card');
                const $topLeft = $('<div>').addClass('top-left').text(shortName);
                const $bottomRight = $('<div>').addClass('bottom-right').text(shortName);
                if (shortName.includes('♠️') || shortName.includes('♣️')) {
                    $topLeft.css('color', 'black');
                    $bottomRight.css('color', 'black');
                } else if (shortName.includes('♥️') || shortName.includes('♦️')) {
                    $topLeft.css('color', 'red');
                    $bottomRight.css('color', 'red');
                }

                $cardDiv.append($topLeft, $bottomRight);
                setTimeout(() => {
                    $dealerCardsContainer.append($cardDiv);
                    setTimeout(() => {
                        $cardDiv.addClass('animate');
                    }, 50);
                }, cardIndex * 150);
            }
        });

        if (data.game_over) {
            console.log(data.result, data.win)
            $result.html('<b>Результат</b>: ' + data.result.join(', ') + '<br><b>Был ли дабл:</b> ' + (data.double_check ? 'да' : 'нет'));
            $win.html('<b>Выигрыш</b>: ' + data.win);

            $.post('/blackjack/end', {win: data.win}, function (response) {
                if (response.success) {
                    $('#balance-int').text(response.new_balance);
                }
            });

            $hit.add($hit2).add($double).add($split).add($stand).add($stand2).prop('disabled', true);

            if (data.result.includes('Игрок')) {
                setTimeout(handlePlayerWin, 1000);
            }
            if (data.result.includes('Блэкджэк у игрока')) {
                setTimeout(handlePlayerWin, 1000);
                setTimeout(handlePlayerWinSmall, 1500);
            }
        } else {
            $result.html('');
            $hit.add($double).add($split).add($stand).prop('disabled', false);

            if (data.player_hands.length > 1) {
                $hit2.add($stand2).show().prop('disabled', false);
            } else {
                $hit2.add($stand2).hide();
            }

            if (data.first_hand_bust || data.first_hand_stand || data.player_scores[0] >= 21) {
                $hit.prop('disabled', true);
                $stand.prop('disabled', true);
            }

            if (data.second_hand_bust || data.second_hand_stand || (data.player_scores.length > 1 && data.player_scores[1] >= 21)) {
                $hit2.prop('disabled', true);
                $stand2.prop('disabled', true);
            }

            if (data.can_split === true) {
                $split.removeClass('btn-outline-info').addClass('btn-info').prop('disabled', !data.can_split);
            } else {
                $split.removeClass('btn-info').addClass('btn-outline-info').prop('disabled', !data.can_split);
            }
        }
        if (data.player_hands[0].length > 2 || (data.player_hands.length > 1 && data.player_hands[1].length > 1)) {
            $split.prop('disabled', true);
            $double.prop('disabled', true);
        } else {
            $split.prop('disabled', !data.can_split);
        }
    }

    $start.click(function () {
        const bet = parseInt($('#bet').val());
        const decksCount = $('#decks-count').val();

        $.get('/check_balance', function (data) {
            if (data.error) {
                alert(data.error);
            } else {
                const balanceInt = parseInt(data.balance);
                if (balanceInt < bet) {
                    alert('Insufficient balance');
                    return;
                }

                const newBalance = balanceInt - bet;
                $('#balance-int').text(newBalance);

                $dealerCardsContainer.empty();
                $playerCardsContainer.empty();
                $playerCardsContainer2.empty();
                existingPlayerCards = [];
                existingDealerCards = [];

                $.post('/blackjack/start', {bet: bet, decks_count: decksCount}, function (data) {
                    if (data.error) {
                        alert(data.error);
                    } else {
                        $split.removeClass('btn-info').addClass('btn-outline-info');
                        updateGame(data);
                        if (!data.game_over) {
                            $hit.add($hit2).add($double).add($stand).add($stand2).prop('disabled', false);
                            $hit2.add($stand2).hide();
                        }
                    }
                });
            }
        });
    });

    $hit.click(function () {
        $.post('/blackjack/hit', function (data) {
            updateGame(data);
        });
    });

    $hit2.click(function () {
        $.post('/blackjack/hit', {hand_index: 1}, function (data) {
            updateGame(data);
        });
    });

    $stand.click(function () {
        $.post('/blackjack/stand', function (data) {
            updateGame(data);
            if (data.first_hand_stand) {
                $hit.prop('disabled', true);
                $stand.prop('disabled', true);
            }
        });
    });

    $stand2.click(function () {
        $.post('/blackjack/stand', {hand_index: 1}, function (data) {
            updateGame(data);
            if (data.second_hand_stand) {
                $hit2.prop('disabled', true);
                $stand2.prop('disabled', true);
            }
        });
    });

    $double.click(function () {
        $.post('/blackjack/double', function (data) {
            updateGame(data);
        });
    });

    $split.click(function () {
        $.post('/blackjack/split', function (data) {
            const firstHandNewCard = data.player_hands[0][1];
            const shortNameFirstHand = cardShortNames[firstHandNewCard] || firstHandNewCard;

            const $firstHandLastCard = $playerCardsContainer.find('.card:last');
            $firstHandLastCard.find('.top-left, .bottom-right').text(shortNameFirstHand);

            if (shortNameFirstHand.includes('♠️') || shortNameFirstHand.includes('♣️')) {
                $firstHandLastCard.find('.top-left').css('color', 'black');
                $firstHandLastCard.find('.bottom-right').css('color', 'black');
            } else if (shortNameFirstHand.includes('♥️') || shortNameFirstHand.includes('♦️')) {
                $firstHandLastCard.find('.top-left').css('color', 'red');
                $firstHandLastCard.find('.bottom-right').css('color', 'red');
            }

            const secondHandFirstCard = data.player_hands[1][0];
            const shortNameSecondHand = cardShortNames[secondHandFirstCard] || secondHandFirstCard;

            const $secondHandCardDiv = $('<div>').addClass('card');
            const $topLeftSecondHand = $('<div>').addClass('top-left').text(shortNameSecondHand);
            const $bottomRightSecondHand = $('<div>').addClass('bottom-right').text(shortNameSecondHand);

            $secondHandCardDiv.append($topLeftSecondHand, $bottomRightSecondHand);
            $('#player-cards-container-2').append($secondHandCardDiv);

            if (shortNameSecondHand.includes('♠️') || shortNameSecondHand.includes('♣️')) {
                $secondHandCardDiv.find('.top-left').css('color', 'black');
                $secondHandCardDiv.find('.bottom-right').css('color', 'black');
            } else if (shortNameSecondHand.includes('♥️') || shortNameSecondHand.includes('♦️')) {
                $secondHandCardDiv.find('.top-left').css('color', 'red');
                $secondHandCardDiv.find('.bottom-right').css('color', 'red');
            }

            $('#player-cards-container-2 .card:first').remove();

            updateGame(data);
        });
    });
})
;

File: static/py/blackjack.py
from pprint import pprint
from typing import List

from itertools import product
from random import shuffle


class Card:
    def __init__(self, suit, name):
        match suit:
            case 'Т':
                self.suit = '♣ трефы'
            case 'Б':
                self.suit = '♦ бубны'
            case 'Ч':
                self.suit = '♥ черви'
            case 'П':
                self.suit = '♠ пики'
            case _:
                self.suit = suit
        match name:
            case '2':
                self.name = '2'
            case '3':
                self.name = '3'
            case '4':
                self.name = '4'
            case '5':
                self.name = '5'
            case '6':
                self.name = '6'
            case '7':
                self.name = '7'
            case '8':
                self.name = '8'
            case '9':
                self.name = '9'
            case 'В':
                self.name = 'Валет'
            case 'Д':
                self.name = 'Дама'
            case 'К':
                self.name = 'Король'
            case 'Т':
                self.name = 'Туз'
            case _:
                self.name = name

    def __str__(self):
        return f'{self.name} {self.suit}'

    def __repr__(self):
        return f'{self.name} {self.suit}'

    def __lt__(self, other):
        if not isinstance(other, Card):
            return NotImplemented
        return CardManager.value(self) < CardManager.value(other)


class Deck:
    def __init__(self, decks_count=8):
        self.cards = [Card(suit, name) for suit, name in
                      product(['Т', 'Б', 'Ч', 'П'],
                              ['2', '3', '4', '5', '6', '7', '8', '9', '10', 'В', 'Д', 'К', 'Т'])] * decks_count
        shuffle(self.cards)

    def __iter__(self):
        return self

    def __next__(self):
        if self.cards:
            return self.cards.pop()
        else:
            raise StopIteration

    def __str__(self):
        return str([str(card) for card in self.cards])

    def __repr__(self):
        return str([str(card) for card in self.cards])


class CardManager:
    @staticmethod
    def value(card: Card) -> int:
        if card.name in ['2', '3', '4', '5', '6', '7', '8', '9', '10']:
            return int(card.name)
        elif card.name in ['Валет', 'Дама', 'Король']:
            return 10
        else:
            return 11

    @staticmethod
    def from_str(card_str: str) -> Card:
        name, suit = card_str.split(' ', 1)
        return Card(suit, name)


class BlackJackGame:
    def __init__(self):
        self.bet = 0
        self.deck = Deck()
        self.dealer_cards: List[Card] = []
        self.player_hands: List[List[Card]] = [[]]
        self.dealer_score = 0
        self.dealer_aces_count = 0
        self.player_scores = [0]
        self.player_aces_count = [0, 0]
        self.double_check = False
        self.game_over = False
        self.first_hand_bust = False
        self.second_hand_bust = False
        self.first_hand_stand = False
        self.second_hand_stand = False
        self.move_count = 0

    def start(self, decks_count=8, bet=0):
        self.bet = bet
        self.move_count = 0
        self.deck = Deck(decks_count)
        self.dealer_cards = [next(self.deck), next(self.deck)]
        self.player_hands = [[next(self.deck), next(self.deck)]]
        self.dealer_score = sum([CardManager.value(card) for card in self.dealer_cards])
        self.dealer_aces_count = 0
        self.player_scores = [CardManager.value(card) for card in self.player_hands[0]]
        self.player_scores = [sum(self.player_scores)]
        self.player_aces_count = [sum(1 for card in self.player_hands[0] if card.name == 'Туз'), 0]
        self.adjust_for_aces(0)
        self.adjust_dealer_for_aces()
        self.game_over = False
        self.check_game_over()

    def hit(self, hand_index=0):
        self.move_count += 1
        if not self.game_over:
            if hand_index >= len(self.player_hands):
                raise IndexError(f"Invalid hand_index: {hand_index}")
            tmp = next(self.deck)
            self.player_hands[hand_index].append(tmp)
            self.player_scores[hand_index] += CardManager.value(tmp)
            if tmp.name == 'Туз':
                self.player_aces_count[hand_index] += 1
            self.adjust_for_aces(hand_index)
            if self.player_scores[hand_index] > 21:
                if hand_index == 0 and len(self.player_hands) > 1:
                    self.first_hand_bust = True
                elif hand_index == 1 and len(self.player_hands) > 1:
                    self.second_hand_bust = True
            self.check_game_over()

    def stand(self, hand_index=0):
        self.move_count += 1
        if not self.game_over:
            if hand_index == 0:
                self.first_hand_stand = True
            if hand_index == 1:
                self.second_hand_stand = True

            if len(self.player_hands) > 1:
                if self.first_hand_stand and self.second_hand_stand:
                    self.dealer_turn()
                else:
                    self.check_game_over()
            else:
                self.dealer_turn()

    def dealer_turn(self):
        while self.dealer_score < 17:
            tmp = next(self.deck)
            self.dealer_cards.append(tmp)
            if tmp.name == 'Туз':
                self.dealer_aces_count += 1
            self.dealer_score += CardManager.value(tmp)
            self.adjust_dealer_for_aces()
        self.game_over = True

    def adjust_for_aces(self, hand_index):
        while self.player_scores[hand_index] > 21 and self.player_aces_count[hand_index] > 0:
            self.player_scores[hand_index] -= 10
            self.player_aces_count[hand_index] -= 1

    def adjust_dealer_for_aces(self):
        while self.dealer_score > 21 and self.dealer_aces_count > 0:
            self.dealer_score -= 10
            self.dealer_aces_count -= 1

    def double(self, hand_index=0):
        self.move_count += 1
        if not self.game_over:
            self.hit(hand_index)
            self.double_check = True
            self.dealer_turn()

    def split(self):
        self.move_count += 1
        if not self.game_over and len(self.player_hands) == 1 and CardManager.value(
                self.player_hands[0][0]) == CardManager.value(self.player_hands[0][1]):
            self.player_hands = [[self.player_hands[0][0]], [self.player_hands[0][1]]]
            self.player_scores = [CardManager.value(self.player_hands[0][0]),
                                  CardManager.value(self.player_hands[1][0])]
            self.hit(0)
            self.hit(1)

    def check_game_over(self):
        if all(score == 21 for score in self.player_scores):
            self.game_over = True
        if len(self.player_hands) > 1:
            if self.first_hand_bust and self.second_hand_bust:
                self.game_over = True

            if self.first_hand_stand and self.second_hand_stand:
                self.dealer_turn()

            if self.first_hand_bust and self.second_hand_stand:
                self.dealer_turn()

            if self.second_hand_bust and self.first_hand_stand:
                self.dealer_turn()

            if self.player_scores[0] == 21 and self.second_hand_bust:
                self.game_over = True

            if self.player_scores[1] == 21 and self.first_hand_bust:
                self.game_over = True

            if self.player_scores[0] == 21 and self.second_hand_stand:
                self.dealer_turn()

            if self.player_scores[1] == 21 and self.first_hand_stand:
                self.dealer_turn()

            if self.player_scores[0] == 21 and self.player_scores[1] == 21:
                self.game_over = True

        elif self.player_scores[0] > 21:
            self.game_over = True

    def get_result(self):
        results = []
        win = 0
        for score in self.player_scores:
            if score > 21:
                results.append('Дилер')
                win += 0
            elif self.dealer_score > 21:
                results.append('Игрок')
                win += self.bet * 2
            elif score == self.dealer_score:
                results.append('Ничья')
                win += self.bet
            elif score == 21:
                results.append('Блэкджэк у игрока')
                win += self.bet * 2.5
            elif self.dealer_score == 21:
                results.append('Блэкджэк у дилера')
                win += 0
            elif score > self.dealer_score:
                results.append('Игрок')
                win += self.bet * 2
            else:
                results.append('Дилер')
                win += 0
        return results, win

    def can_split(self):
        return (not self.game_over and len(self.player_hands) == 1 and
                CardManager.value(self.player_hands[0][0]) == CardManager.value(self.player_hands[0][1]))

    def to_dict(self):
        return {
            'bet': self.bet,
            'dealer_cards': [str(card) for card in self.dealer_cards],
            'player_hands': [[str(card) for card in hand] for hand in self.player_hands],
            'dealer_score': self.dealer_score,
            'player_scores': self.player_scores,
            'double_check': self.double_check,
            'game_over': self.game_over,
            'first_hand_bust': self.first_hand_bust,
            'second_hand_bust': self.second_hand_bust,
            'first_hand_stand': self.first_hand_stand,
            'second_hand_stand': self.second_hand_stand,
            'result': self.get_result()[0] if self.game_over else None,
            'win': self.get_result()[1] if self.game_over else None,
            'can_split': self.can_split(),
            'move_count': self.move_count
        }

    @classmethod
    def from_dict(cls, data):
        game = cls()
        game.bet = data['bet']
        game.dealer_cards = [CardManager.from_str(card) for card in data['dealer_cards']]
        game.player_hands = [[CardManager.from_str(card) for card in hand] for hand in data['player_hands']]
        game.dealer_score = data['dealer_score']
        game.player_scores = data['player_scores']
        game.double_check = data['double_check']
        game.game_over = data['game_over']
        game.first_hand_bust = data['first_hand_bust']
        game.second_hand_bust = data['second_hand_bust']
        game.first_hand_stand = data['first_hand_stand']
        game.second_hand_stand = data['second_hand_stand']
        game.move_count = data['move_count']
        return game
